### PR TITLE
SDI-357 unknown patients API

### PR DESCRIPTION
**Pull request by mikehalmamoj**
And switch to internal discharge-to-hospital service

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsrestrictedpatientsapi/controllers/RestrictedPatientsBatchController.kt [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+  @Hidden
+  @PostMapping(value = ["/process-unknown-patients"])
+  @PreAuthorize("hasRole('RESTRICTED_PATIENT_MIGRATION')")
+  /**
+   * Internal endpoint to migrate unknown patients into Nomis and add them to Restricted Patients
+   * TODO SDI-357 This is still a WIP
+   */
+  fun processUnknownPatients(@RequestBody patients: List<String>): List<UnknownPatientResult> =
+    unknownPatientService.migrateInUnknownPatients(patients)
+}
```
Comment (mikehalmamoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsrestrictedpatientsapi/controllers/RestrictedPatientsBatchController.kt:36]: This is new - the rest is from renaming the class

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsrestrictedpatientsapi/services/RestrictedPatientsService.kt [hunk 2]
```diff
@@ -50,7 +50,7 @@ class RestrictedPatientsService(
       fromLocationId = dischargeToHospitalWithDefaultSupportingPrison.fromLocationId,
       hospitalLocationCode = dischargeToHospitalWithDefaultSupportingPrison.hospitalLocationCode,
       supportingPrisonId = dischargeToHospitalWithDefaultSupportingPrison.supportingPrisonId!!,
-      dischargeTime = LocalDateTime.now(clock),
+      dischargeTime = dischargeToHospital.dischargeTime ?: LocalDateTime.now(clock),
```
Comment (mikehalmamoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsrestrictedpatientsapi/services/RestrictedPatientsService.kt:53]: Allow overrides to the movement time - needed for these historical movements

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsrestrictedpatientsapi/services/UnknownPatientService.kt [hunk 3]
```diff
@@ -55,8 +58,15 @@ class UnknownPatientService(private val agencyFinder: AgencyFinder, private val
...
-        .let { prisonApiGateway.dischargeToHospital(it) }
+      DischargeToHospitalRequest(
+        offenderNumber,
+        "Historical hospital release added to NOMIS for addition to Restricted Patients",
+        prisonCode,
+        hospitalCode,
+        prisonCode,
+        hospitalOrderDate.atStartOfDay(),
+      )
+        .let { restrictedPatientsService.dischargeToHospital(it) }
```
Comment (mikehalmamoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsrestrictedpatientsapi/services/UnknownPatientService.kt:69]: Call the internal service - I've realised it does everything we need.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsrestrictedpatientsapi/integration/ProcessUnknownPatientsIntTest.kt [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+import org.junit.jupiter.api.Test
+import org.mockito.ArgumentMatchers.anyList
+import org.mockito.kotlin.whenever
+import org.springframework.test.context.ActiveProfiles
+
+@ActiveProfiles("test")
+class ProcessUnknownPatientsIntTest : IntegrationTestBase() {
+
+  @Nested
+  inner class Security {
```
Comment (mikehalmamoj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsrestrictedpatientsapi/integration/ProcessUnknownPatientsIntTest.kt:13]: Started with tests to secure the endpoint. More integration tests to come for the functionality.

**Review by steverendell — APPROVED** [review]
lgtm

**Pull request merged**